libsignal-service/src/gv2/operations.rs

```diff
@@ -0,0 +1,221 @@
+            uuid: uuid.to_vec(),
+            profile_key: bincode::serialize(&profile_key)?,
+            role: member.role,
+            joined_at_revision: member.joined_at_revision,
+        })
+    }
+
+    fn decrypt_pending_member(
+        &self,
+        member: EncryptedPendingMember,
+    ) -> Result<DecryptedPendingMember, GroupDecryptionError> {
+        let inner_member =
+            member.member.ok_or(GroupDecryptionError::WrongBlob)?;
+        let uuid = self.decrypt_uuid(&inner_member.user_id).unwrap_or_default();
+        let added_by = self.decrypt_uuid(&member.added_by_user_id)?;
+
+        Ok(DecryptedPendingMember {
+            uuid: uuid.to_vec(),
+            role: inner_member.role,
+            added_by_uuid: added_by.to_vec(),
+            timestamp: member.timestamp,
+            uuid_cipher_text: inner_member.user_id,
+        })
+    }
+
+    fn decrypt_requesting_member(
+        &self,
+        member: EncryptedRequestingMember,
+    ) -> Result<DecryptedRequestingMember, GroupDecryptionError> {
+        let (uuid, profile_key) = if member.presentation.is_empty() {
+            let uuid = self.decrypt_uuid(&member.user_id)?;
+            let profile_key =
+                self.decrypt_profile_key(&member.profile_key, uuid)?;
+            (uuid, profile_key)
+        } else {
+            let profile_key_credential_presentation: ProfileKeyCredentialPresentation = bincode::deserialize(&member.presentation)?;
+            let uuid = self.group_secret_params.decrypt_uuid(
+                profile_key_credential_presentation.get_uuid_ciphertext(),
+            )?;
+            let profile_key = self.group_secret_params.decrypt_profile_key(
+                profile_key_credential_presentation
+                    .get_profile_key_ciphertext(),
+                uuid,
+            )?;
+            (uuid, profile_key)
+        };
+        Ok(DecryptedRequestingMember {
+            profile_key: bincode::serialize(&profile_key)?,
+            uuid: uuid.to_vec(),
+            timestamp: member.timestamp,
+        })
+    }
+
+    fn decrypt_blob(&self, bytes: &[u8]) -> GroupAttributeBlob {
+        if bytes.is_empty() {
+            GroupAttributeBlob::default()
+        } else if bytes.len() < 29 {
+            log::warn!("bad encrypted blob length");
+            GroupAttributeBlob::default()
+        } else {
+            use group_attribute_blob::Content;
+            self.group_secret_params
+                .decrypt_blob(bytes)
+                .map_err(|_| GroupDecryptionError::ZkGroupError)
+                .and_then(|b| {
+                    GroupAttributeBlob::decode(Bytes::copy_from_slice(&b[4..]))
+                        .map_err(GroupDecryptionError::ProtobufDecodeError)
+                })
+                .unwrap_or_else(|e| {
+                    log::warn!("bad encrypted blob: {}", e);
+                    GroupAttributeBlob::default()
+                })
+        }
+    }
+
+    fn decrypt_title(
```

Comment:
I'd really like to avoid macros as much as we can, as it makes browsing the code just... painful. What do you mean exactly by scaffold? One thing we might want to do is directly `impl` things on the protobuf mirrored structs themselves (since it's included in the crate, we should be able to.)

Reply:
It's just all the same code with different function names, that's why I suggested a macro. It can probably be done with enum and visitor too, but don't feel pressured into refactoring this yet. We can do it later.

Reply:
Hmm ok, I'd say that if the code duplicates more than x2 (it's currently exactly 2) we should look into a macro. I think it'll only show when we implement other protobuf structs and I'd do it then.